### ISSUE TITLE
Rollback being called multiple times in CANCEL

### ISSUE THREAD

**Issue by RachelXGanon** [oleg-shilo/wixsharp#1435]
Hi,
I'd like to understand a bit more how the major upgrade works.
Using ` project.MajorUpgrade.Schedule = UpgradeSchedule.afterInstallInitialize;`

Does the upgrade remove old version and afterwards install new version? 
And what occur if I CANCEL the upgrade?

I defined the following CA:
```
  //install manifest
              new ElevatedManagedAction(CustomActions.InstallManifest, Return.check, When.After,
                        Step.InstallFiles, Condition.NOT_Installed, Sequence.InstallExecuteSequence)
               {Execute = Execute.deferred, Impersonate = false },
               
              new ElevatedManagedAction(CustomActions.UninstallManifest, Return.check, When.Before,
                        Step.RemoveFiles, Condition.BeingUninstalled, Sequence.InstallExecuteSequence)
               {Execute = Execute.deferred, Impersonate = false },
               
  //rollback
               new ElevatedManagedAction(CustomActions.UninstallManifest, Return.check, When.After,
                        Step.InstallFiles, Condition.NOT_Installed, Sequence.InstallExecuteSequence)
               {Execute = Execute.rollback, Impersonate = false },

                new ElevatedManagedAction(CustomActions.InstallManifest, Return.check, When.After,
                        Step.RemoveFiles, Condition.BeingUninstalled, Sequence.InstallExecuteSequence)
               {Execute = Execute.rollback, Impersonate = false }
```

CANCEL trigger the rollback, but as installation done I see the manifest is not registered.

I checked the logs and it call `UninstallManifest` after `InstallManifest`.
How does it work?
 I thought CANCEL suppose to trigger rollback and uninstall the manifest(done)
 and re install the previous version with `InstallManifest` (not done, or done in the wrong order..)

**Comment by RachelXGanon** [comment]
Ok, I updated the code into:
 ```
//install manifest
              new ElevatedManagedAction(CustomActions.InstallManifest, Return.check, When.After,
                        Step.InstallFiles, Condition.NOT_Installed, Sequence.InstallExecuteSequence)
               {Execute = Execute.deferred, Impersonate = false },
           
               
  //rollback
               new ElevatedManagedAction(CustomActions.UninstallManifest, Return.check, When.After,
                        Step.InstallFiles, Condition.NOT_Installed, Sequence.InstallExecuteSequence)
               {Execute = Execute.rollback, Impersonate = false },

                new ElevatedManagedAction(CustomActions.InstallManifest, Return.check, When.After,
                        Step.RemoveFiles, Condition.BeingUninstalled, Sequence.InstallExecuteSequence)
               {Execute = Execute.rollback, Impersonate = false }
```

So, `UninstallManifest` should be called only once in the rollback.
Checking the logs in called twice:
First time in the RemoveExitingProducts - which is good.
Second time in the end of the installation -
```
ActionStart: Action 11:38:51: Rollback. Rolling back action:
ActionData: CancelRequestHandler
ActionData: Publishing product information
ActionData: Publishing Product Features
ActionData: Registering product
ActionData: Starting services
ActionData: MsiConfigureServices
ActionData: ExecServiceConfig
ActionData: RollbackServiceConfig
ActionData: Installing new services
ActionData: WixSharp_AfterInstall_Action
ActionData: InstallManifest
ActionData: Copying new files
ActionData: UninstallManifest
Info: SFXCA: Extracting custom action to temporary directory: C:\WINDOWS\Installer\MSI8B31.tmp-\
Info: SFXCA: Binding to CLR version v4.0.30319
Info: Calling custom action Installer!CustomActions.UninstallManifest
```
Any idea why it being called twice? Is there a rollback twice?

**Comment by oleg-shilo** [comment]
> Does the upgrade remove old version and afterwards install new version?

Yes

>And what occur if I CANCEL the upgrade

If you do it after the old version is removed, you will have no product installed.

**Comment by RachelXGanon** [comment]
Weird, I do have the old version installed.
Anyway, how can Wixsharp support at rollback of upgrade?

**Comment by oleg-shilo** [comment]
> Weird, I do have the old version installed.

Yes, it could be the case if you manage to cancel the installation before it starts. Still, you may want to verify my answers as I can be wrong. The easiest way to do this is not thrifting through the documentation but by simply testing your scenario with reliable MSIs. IE you can do it with the "Major Upgrade" sample

> ...how can Wixsharp support rollback of upgrade, reinstall old version?
WixSharp cannot. The rollback/upgrade is implemented by MSI. WiX is implementing building the right MSI. WixSharp is only playing the role of the glue between these two in your scenario.

If I was in your shoes I would stop there and not try to overcome some MSI architectural constraints. 
You have done good job bu building a proper, upgradable MSI. You diligently check if the target environment is compatible with the product version you are installing. If your user was not careful enough and started clicking all buttons for no reason then you are not responsible for clearing it. BUt you are responsible for letting the user know what happens and how to proceed in the new situation. IE "Just retry installing the product", or "Uninstall the old product manually and install a new one.".

**Comment by RachelXGanon** [comment]
Got it, thank you Oleg.
